### DISCUSSION TITLE
APIとして欲しい機能

### DISCUSSION THREAD

**Discussion by nozaki-pfr** [pf-robotics/kachaka-api#23] [Ideas]
＋押せるように１機能１コメントで書きます。
これまでに要望を聞いたコメントをまずは書いておきます

**Comment by nozaki-pfr** [comment]
バックカメラの画像

**Reply by nozaki-pfr** [reply]
#64 でリリースされました

**Comment by nozaki-pfr** [comment]
任意の音の再生

**Comment by asagin** [comment]
Arduinoなどから簡単にカチャカを操作するために、gRPCのAPIをラップしてHTTPでAPIをコールできる仕組みを用意してもらえると嬉しいです。

**Reply by nozaki-pfr** [reply]
kachaka_apiライブラリをPOST, GETで実行するデモを作成してみました https://github.com/pf-robotics/kachaka-api/blob/main/python/demos/url_kachaka_api.ipynb

**Reply by nozaki-pfr** [reply]
動作確認時の環境のIPアドレスが残ってしまってましたので修正中です:bow: https://github.com/pf-robotics/kachaka-api/pull/31

**Reply by nozaki-pfr** [reply]
治りました！

**Comment by akinobu1998** [comment]
任意の発話を聞き取る機能．スマートスピーカーを使わずに，ボイスコマンド一覧にないコマンドも認識させたいです．

**Comment by youtalk** [comment]
ToF（3Dセンサー）の画像

**Reply by kmurase-pfr** [reply]
https://github.com/pf-robotics/kachaka-api/pull/87 でリリースされました

**Comment by nozaki-pfr** [comment]
エラーの取得

**Reply by ysuzuysuzu** [reply]
エラー取得時にコマンドをキャンセル、別のコマンドを実行などを行いたいのですが、
こちらは実装のめどは立っているのでしょうか

**Reply by inabas-pfr** [reply]
@ysuzuysuzu 
お待たせし大変申し訳ございません。
v3.1.10 のリリースに下記の対応が入りました。
#95 
> GetRobotErrorCodeJson: エラーコード一覧のjsonを取得 (GetErrorの返り値のerror_codeのため)
> GetError: 発生している非同期エラーを取得

ご要望頂いている内容がこちらで問題がないかご確認いただけますと幸いです。

**Comment by nozaki-pfr** [comment]
バッテリ情報の取得

**Reply by kmurase-pfr** [reply]
https://github.com/pf-robotics/kachaka-api/pull/87 でリリースされました

**Reply by yoshikai-qt** [reply]
@kmurase-pfr 機能追加ありがとうございます。早速アップデートして確認しようとしていますが、現状、python apiから直接grpcを叩いて取得する方法のみサポートされているということでよろしいでしょうか。こちらについて、ROS2のインタフェースに対応する計画はございますでしょうか。どうぞ、よろしくお願い致します。

**Reply by kmurase-pfr** [reply]
ROS2ブリッジとしての実装はまだ行っておりませんが、検討します。

grpc_ros2_bridgeのソースコードとDockerfileは公開しておりますので、ご自身で追加頂くこともできると思います。
https://github.com/pf-robotics/kachaka-api/blob/main/Dockerfile.ros2
https://github.com/pf-robotics/kachaka-api/tree/main/ros2/kachaka_grpc_ros2_bridge/

**Reply by yoshikai-qt** [reply]
@kmurase-pfr ありがとうございます。こちらについても承知致しました。トライしてみます。

**Reply by yoshikai-qt** [reply]
@kmurase-pfr とりあえず対応してみたものができました。トピック名の場所などポリシーはあるとお思いますが、PRを作ってみましたので、よろしくお願い致します。
https://github.com/pf-robotics/kachaka-api/pull/89

**Comment by yoshikai-qt** [comment]
現在、ROS2のインタフェースを使ってアプリケーション開発を進めておりますが、
Lidarデータのpublishについて要望があります。現状では、/kachaka/manual_control/cmd_velに
何らかの入力がない限りは/kachaka/lidar/scanの出力が出ない仕様になっていると思いますが、
ユーザ側でリスクをとって常時/kachaka/lidar/scanを出力するAPIを用意して頂けないでしょうか。
動かさないが観測だけしたいという状況はそれなりにありまして、Lidar寿命に影響するのは重々承知
しておりますが、必要に応じて観測のみを行えるようになるとありがたいです。ご検討よろしくおねがいします。

**Reply by kmurase-pfr** [reply]
SW3.0で追加予定です！

**Reply by kmurase-pfr** [reply]
https://github.com/pf-robotics/kachaka-api/pull/87 でリリースされました
使い方は以下をご覧下さい
https://github.com/pf-robotics/kachaka-api/blob/main/python/demos/get_laser_scan.ipynb

**Reply by yoshikai-qt** [reply]
@kmurase-pfr ありがとうございます！ちなみに、現状はpythonのAPIから直接操作する形のみのサポートでしょうか。ROS2インタフェース（トピック、サービス等）からLaserScanActivator.activate()を叩く機能のサポートは今後計画されておりますでしょうか。

**Reply by kmurase-pfr** [reply]
ROS2ブリッジとしての実装はまだ行っておりませんが、検討します。

grpc_ros2_bridgeのソースコードとDockerfileは公開しておりますので、ご自身で追加頂くこともできると思います。
https://github.com/pf-robotics/kachaka-api/blob/main/Dockerfile.ros2
https://github.com/pf-robotics/kachaka-api/tree/main/ros2/kachaka_grpc_ros2_bridge/

**Reply by yoshikai-qt** [reply]
@kmurase-pfr ありがとうございます。承知致しました。

**Comment by nozaki-pfr** [comment]
global localization（落ち着いて）の実行

**Comment by nozaki-pfr** [comment]
ショートカットの実行

**Reply by inabas-pfr** [reply]
v3.1.10 のリリースで対応済みです。
#95 
> GetShortcuts: スマホアプリのホーム画面にあるショートカットの一覧を取得
> StartShortcutCommand: ショートカットを実行

**Comment by nakayamayotetsu-pfr** [comment]
APIで作った機能をアプリ（ショートカットなど）からトリガーしたい

**Comment by youtalk-pfr** [comment]
ロケーションの新規登録、編集（ユーザフィードバック）

**Comment by h-wata** [comment]
get_png_mapのオプションで、マップ名だけ取得したい。

**Reply by h-wata** [reply]
https://github.com/pf-robotics/kachaka-api/pull/77 にて同等のAPIが実装されたので取り下げます。

https://github.com/pf-robotics/kachaka-api/blob/cc762afd96f6d35731367b7096f6a1e5a7335706/python/kachaka_api/base.py#L426-L436

**Reply by watanabe-pfr** [reply]
こちら回答遅れましたが、そうですね、get_current_map_idで現在の地図が取得できるようになったため、こちらご利用ください。

**Comment by h-wata** [comment]
Shelfとドッキングしているかどうか。※すでにあったらすみません。

**Reply by watanabe-pfr** [reply]
SWv2.7で追加予定です！

**Reply by watanabe-pfr** [reply]
https://github.com/pf-robotics/kachaka-api/releases/tag/v2.7.8 2.7.8をリリースし、こちら追加されました！

**Comment by h-wata** [comment]
Mapの切り替えAPIは用意予定ございますでしょうか？

**Reply by watanabe-pfr** [reply]
SW v2.7ではサポートできませんが、近いバージョンで対応を予定しております。
ちなみに、どういった使い方を想定されていますか？(仕様の参考にしたく)

**Reply by h-wata** [reply]
コメントありがとうございます。
Opern-RMFのfleet_adapterを開発しておりまして、エレベータ連携も検討しております。
そのため、エレベータで別フロアに移動した際に1F -> 2Fへのマップの切り替えを行いたいと考えています。
引数で`map_id(str)`と`initial_pose(x,y,theta)`を与えると、マップが切り替わり自律移動が可能な状況になると嬉しく感じております。

**Reply by watanabe-pfr** [reply]
ありがとうございます、大変参考になります…！(すごく楽しそう)

> 引数でmap_id(str)とinitial_pose(x,y,theta)を与えると、マップが切り替わり自律移動が可能な状況になると嬉しく感じております。

こちらすごく良いと思うので、基本これができるように進めようと思います。

**Reply by sega23** [reply]
エレベータ連携のために、是非ともほしいところです。

**Reply by kmurase-pfr** [reply]
https://github.com/pf-robotics/kachaka-api/pull/87 でリリースされました

**Reply by h-wata** [reply]
ありがとうございます！

**Reply by h-wata** [reply]
~~※Issueにすべきかもしれませんが、関連なのでこちらで失礼します。もしもIssueのほうが良ければご指示ください。
grpcに対する理解が乏しく、SwitchMapの使い方がわかりません。
`python/kachaka-api/base.py`にて下記のような実装を行いPlaygroundにて試してみました。~~

```python
    def switch_map(self, map_id: str, initial_pose: pb2.Pose) -> pb2.Result:
        request = pb2.SwitchMapRequest(
            map_id=map_id, initial_pose=initial_pose)
        response: pb2.SwitchMapResponse = self.stub.SwitchMap(request)
        return response.result
```
結果として、メソッドがない旨のエラーが出ました。
```
File [~/kachaka-api/python/kachaka_api/base.py:535](http://192.168.1.100:26501/lab/tree/kachaka-api/python/kachaka-api/python/kachaka_api/base.py#line=534), in KachakaApiClientBase.switch_map(self, map_id, initial_pose)
    534 def switch_map(self, map_id: str, initial_pose: pb2.Pose) -> pb2.Result:
--> 535     request = pb2.SwitchMapRequest(
    536         map_id=map_id, initial_pose=initial_pose)
    537     response: pb2.SwitchMapResponse = self.stub.SwitchMap(request)
    538     return response.result

AttributeError: module 'kachaka_api.generated.kachaka_api_pb2' has no attribute 'SwitchMapRequest'
```
~~確かに ['kachaka_api.generated.kachaka_api_pb2'](https://github.com/pf-robotics/kachaka-api/blob/main/python/kachaka_api/generated/kachaka_api_pb2_grpc.py)を見ると、未定義のようなのですが現状こちらからの利用は想定されていないでしょうか？
また、利用するためにはどのような実装が必要でしょうか。demoフォルダに利用例などがあれば嬉しいです。
ご検討よろしくおねがいします。~~


[ドキュメント](https://github.com/pf-robotics/kachaka-api/blob/main/docs/GRPC.md)を再度見直し自己解決しました。
お騒がせしました。

**Reply by watanabe-pfr** [reply]
protoを更新したら、pythonのコード生成もし直さないといけないんですよね。それが漏れていたので、後で修正します。
お手元でも、tools/update_kachaka_api_系のスクリプトを叩くことで更新していただけるので、よかったらPR投げてくださるととても助かります！

**Reply by h-wata** [reply]
https://github.com/pf-robotics/kachaka-api/pull/90

こちらにてPR投げさせていただきました。Reviewいただければ幸いです。

**Comment by akinobu1998** [comment]
英語を話す機能

**Comment by MasakazuKONDO** [comment]
音量の調整

**Reply by inabas-pfr** [reply]
v3.1.10 のリリースに下記の対応が入りました。
#95 
> GetSpeakerVolume: 音量の取得(0-10)
> SetSpeakerVolume: 音量の設定(0-10)

ご確認のほどよろしくお願いいたします。

**Comment by yutaroha** [comment]
Open-RMFを用いてカチャカプロを複数台制御を行っています。
その際に、障害物など検知した際にその振る舞いをカスタマイズできるような機能があると嬉しいです。

## ほしい機能のイメージ
- 工場など生産現場で使用する場合に回避してよけると危険な場合があり、一定範囲内に人がいたら減速する、停止するなどができると嬉しいです。人以外にもフリート制御で認識できていない有軌道のAGVが走っているケースもあり、衝突防止のためにこのような機能が必要と考えています。
- フリート制御側のプログラムでgRPCのAPIを用いて、物体検出情報を用いて止めることもできるかとは思いますが、カチャカとフリート制御側の通信断が発生した場合に動かなくなってしまうのでできればカチャカ内でそのような動作ができると嬉しいと考えています。
- ビヘイビアツリーを渡して、このケースであれば減速、このケースであれば停止するなどが可能であるとよりフレキシブルに動作変更可能であると考えています。(ノーコード・ローコードなUIで変更できたり、ビヘイビアツリーをgRPCのAPIで渡せて動作を変更可能なイメージ)

**Reply by Akinori1025** [reply]
同じく　API制御で回避動作するよりも減速　及び停止機能はほしいです

**Comment by yutaroha** [comment]
シェルフを登録する機能が欲しいです。

背景として、複数台のカチャカでシェルフを共有する場合にカチャカの台数毎にシェルフの登録を行う必要があるため。
例えば、10台のカチャカで1つのシェルフを使用する場合、10台それぞれでシェルフの登録が必要となる。
そのため、シェルフのID？を引数にしてシェルフを登録できる関数があると嬉しいです。
(シェルフの位置はこちらで管理しており、DockShelfCommandで前にあるシェルフをドックさせるという形で使用しています。)

この機能は複数台のユースケースのみで使用することになると思うので、本APIに入れるのは違うのかもしれませんが。。
登録していないシェルフでも運ぶことができる形でも問題ないです。

**Reply by watanabe-pfr** [reply]
地図のインポート・エクスポートの機能を使っていただくと、1台分登録を済ませた後に他の機体にコピーすればこの登録の手間の問題は解決しそうですが、いかがでしょうか？

**Reply by yutaroha** [reply]
ご回答ありがとうございます。

上記の形で問題なさそうです。ありがとうございます。

前にインポートしたときに地図だけインポートされてシェルフの情報が入っておらず、地図情報だけ入るものだと誤認しておりました。
本日確認したところ、シェルフの情報もインポートできていそうでした。

ありがとうございました。

**Comment by yutaroha** [comment]
カチャカアプリについてのほしい機能となります。
リクエストする場所が適切で無いようであれば申し訳ありません。

カチャカアプリでカチャカプロを接続して使用しています。
**LAN接続のみで使用する**をONで接続しています。

アプリケーションでカチャカプロ接続先を1つのIPアドレスしか指定できないため、**カチャカプロ毎にIPアドレスを保持できるように**なると嬉しいです。

現在はカチャカプロ1に接続し、その後カチャカプロ2に接続する際にアプリのその他設定からIPアドレス編集後にカチャカプロの選択を切り替える必要があるため。

**Reply by inabas-pfr** [reply]
根本的な解決にはならないのですが、、「kachaka レストラン」のアプリをリリースしております。
アプリをふたつ入れることで、アプリの使い分けで設定を変更する手間を減らすことが可能です。
- カチャカプロ1：「kachaka」アプリを使う
- カチャカプロ2：「kachaka レストラン」アプリを使う
　　★ただしレストランアプリは棚の運搬の指定はできず、あくまで移動を目的としたアプリです。

条件に合う場合はご利用いただけますと幸いです。
- https://kachaka.zendesk.com/hc/ja/articles/6966050952847
　　※詳細については「Kachakaレストラン手順書_vX.X.X.pdf」をご覧ください。

また本件については社内でも検討しております。

**Reply by yutaroha** [reply]
ありがとうございます。
ぜひ使ってみたいと思います。

そこまで大きくない飲食店であれば、このアプリとカチャカだけでロボット導入できてしまいそうですね！

**Comment by yutaroha** [comment]
最大速度を指定して移動ができるようになるとうれしいです。
危険な場所では低速で、見通しの良い場合は早いスピードにするなど。

MoveToPoseCommandで最大速度が指定可能であると上位のアプリ側から指示できるのでパラメータに追加されるとうれしいです。

**Reply by masayukisobe-pfr** [reply]
カチャカ Pro モデル限定となりますが、カチャカ ソフトウェア ver 3.8.5 より、走行速度エリア機能が追加されました。
APIではなくアプリからの設定となりますが、ご指摘のユースケースに使えるかもしれません。
https://kachaka.zendesk.com/hc/ja/articles/11792437138319--Kachaka-Pro-%E8%B5%B0%E8%A1%8C%E9%80%9F%E5%BA%A6%E3%82%A8%E3%83%AA%E3%82%A2%E3%82%92%E6%8C%87%E5%AE%9A%E3%81%99%E3%82%8B

**Reply by yutaroha** [reply]
ありがとうございます。
早速試してみます！

**Reply by yutaroha** [reply]
走行速度エリア機能についてアプリで設定を試してみました。
ＵＩ部分になりますが、タブレットでの操作時に走行速度設定が若干反応が悪かったので情報共有させていただきます。
スライダを押してるつもりなのですが、うまく反応しないことがありました。(スクロールが反応してる？)

https://github.com/user-attachments/assets/63a0a321-a903-427c-855f-523d4cbc250c

**確認機種**
OS: Android
モデル：Lenovo TB-X306F
ディスプレイサイズ：10.1型

※ iPhone SE3で操作した場合は本現象は発生していませんでした。

**Reply by yutaroha** [reply]
走行速度エリア機能を使用して、走行速度の指定をしたのですが特に挙動が変わりませんでした。
何か必要なオプション設定などありますでしょうか。

**動作環境**
カチャカ：カチャカプロ
カチャカバージョン：v.3.8.5
カチャカアプリ(Android)：2.20.7

**確認内容**
- カチャカアプリにて、走行速度エリア設定(動く遅さを最も遅く)
- 走行速度エリア設定をした場所をカチャカプロが通貨するように、行先指定
- 走行速度エリア設定を行ったエリア内に入っても速度が変化せず

**その他**
- マップを新たに作成して、走行速度エリア設定をしても特に効果がありませんでした。
- iOSでも念のため確認しましたが変わりませんでした

**Reply by inabas-pfr** [reply]
@yutaroha 詳細な情報のご提供誠にありがとうございます。
急ぎ、社内に共有を進めさせていただきます。

**Reply by inabas-pfr** [reply]
@yutaroha お試しいただきありがとうございます。

特に特殊な操作は必要はありません、、
こちらのマニュアルの最後の「OK」をタップし、保存操作が行えているかを念のためご確認いただけますでしょうか？
**[Kachaka Pro]走行速度エリアを指定する**
https://kachaka.zendesk.com/hc/ja/articles/11792437138319
> 8.「OK」をタップし設定を保存します

**Reply by yutaroha** [reply]
@inabas-pfr 
ありがとうございます。OKボタンはタップして保存しております。

1つ気になるのは、走行エリア設定後にマップを開くのですが走行エリアの表示はされておりませんでした。
(以下の動画にはありませんが、進入禁止エリアは表示されていました)

https://github.com/user-attachments/assets/3dfed412-e41c-4573-96ac-27ab85e4142f

**Reply by inabas-pfr** [reply]
ご確認いただきありがとうございます。
家具を載せた状態で走行速度の変化が見られるかをご確認いただけますでしょうか？

また通常のカチャカの速度を「速い」に設定いただくと、
「遅い」に設定したエリアとの走行の差分動作がわかりやすくなります。
https://kachaka.zendesk.com/hc/ja/articles/6362721702543
https://kachaka.zendesk.com/hc/ja/articles/8764251540367
　※速度は「走行速度エリア」と「カチャカ設定」と差分はありません。

**Reply by inabas-pfr** [reply]
> 1つ気になるのは、走行エリア設定後にマップを開くのですが走行エリアの表示はされておりませんでした。

こちらのUI表示は想定している挙動となります。
現時点では、「走行速度エリア指定」の画面でのみ設定情報が閲覧可能です。

**Reply by inabas-pfr** [reply]
五月雨式にすみません。。

> スライダを押してるつもりなのですが、うまく反応しないことがありました。(スクロールが反応してる？)

こちらの不具合については、次回(2月中旬以降)のv2.21のスマホアプリで修正を予定しております。
ご不便をおかけしますが、何卒よろしくお願いいたします。

**Reply by yutaroha** [reply]
@inabas-pfr 

> こちらの不具合については、次回(2月中旬以降)のv2.21のスマホアプリで修正を予定しております。
> ご不便をおかけしますが、何卒よろしくお願いいたします。

ありがとうございます。承知いたしました。

> 家具を載せた状態で走行速度の変化が見られるかをご確認いただけますでしょうか？

こちら確認したところ、特に変化が見られませんでした。

新たにわかりましたことがありますので展開いたします。

**走行速度エリア指定を1つおくと、減速する**ことを確認しました。
**走行速度エリア指定を2つおくと、制限されない**ことを確認しました。
**複数の走行速度エリア指定を置くことで本現象が発生する可能性がある**ように見受けられます。

走行速度エリア指定を2つ置いた後に1つを削除すると、正しく減速することを確認しました。
このとき、2つの走行速度エリアは重ならないように設定しております。

**Reply by inabas-pfr** [reply]
レポートいただきありがとうございます。
社内にも共有し、確認を進めてまいります。

**Reply by inabas-pfr** [reply]
@yutaroha 
ご不便をおかけし大変申し訳ございませんでした。
こちらの修正が本日リリースのv3.9.5に入りましたので、ご確認いただけますと幸いです。

**Reply by yutaroha** [reply]
@inabas-pfr 
アップデート後確認したところ無事動作いたしました。
ありがとうございます。

**Reply by yutaroha** [reply]
使用していて感じた点としましては、マップ画面で走行速度エリアの設定も見れるとうれしいと感じました。
マップ画面で各種レイヤ(進入禁止エリア、進入可能エリア、走行速度エリア)の可視化ON/OFFのようなものがあると確認しやすいと思いました。
ご検討いただけましたら幸いです。

**Comment by yutaroha** [comment]
走行音を常時ON/OFFできるような機能が欲しいです。
工場内のルールによっては、作業者にAMRの存在を知らせるために、止まっている最中でも音を鳴らしておく必要があるため。

**Comment by s-remi03** [comment]
シェルフのidを指定したらそのシェルフとドッキングするのAPIがほしいです。既にありましたら申し訳ございません。

**Reply by inabas-pfr** [reply]
カチャカAPIでご提供している類似のAPIは「家具を目的地に移動」、「前方の家具を載せる」などです。
ご要望のAPIは目的地への移動はせずに、シェルフにドッキングしてその場にとどまるということでしょうか？

https://github.com/pf-robotics/kachaka-api/blob/main/docs/kachaka_api_client.ipynb
>**家具の移動**
>client.move_shelf("S01", "L01")

>家具を載せる
>client.dock_shelf()

**Reply by s-remi03** [reply]
ご返信ありがとうございます。ご返信の通り、IDを指定するとドッキングのみ行いその場に留まる機能が欲しいなと思っています。
具体的な活用例としては、
・シェルフとドッキング後にmove_to_pose()関数を使用して登録されていない場所に移動
→登録されていない場所にあるシェルフを選択してドッキング
などを考えています。

「前方の家具を載せる」APIでは、家具の前にいかないと使えない為、ID指定で自動的にドッキングできると嬉しいです。

**Comment by yutaroha** [comment]
ナビゲーション時に、たどり着けない場合に5分程度周りを見渡す動きをした後に、停止する流れのように見受けられます。
自動で点検業務で巡回する場合に5分だと長すぎる場合があるため、タイムアウトを設けられるようになるとうれしいです。

取り急ぎは、位置を定期的に取得して、一定時間内に一定距離以上動いてなければキャンセルするなどの対応にて進めようと考えております。

**Reply by inabas-pfr** [reply]
@yutaroha  社内にもFBは進めております。
追加の対策として下記の「位置の調整を省く」の設定をご案内させてください。
- [マニュアル：目的地に移動後の位置と方向の調整を省く](https://kachaka.zendesk.com/hc/ja/articles/9547487632783)
  - ★UIがアップデートされているためマニュアルは更新予定です

ただし設定すると停止位置精度が下がってしまうため、現場によっては解決策にならない可能性もございます。。
まずはお試し頂けますと幸いです。

**Reply by yutaroha** [reply]
@inabas-pfr 
ありがとうございます。

カチャカの移動問題というよりは、マッピングしたときにはなかった障害物がナビゲーションのゴール地点に配置されて、たどり着けずにタイムアウトを待って次の地点に進むという流れになってしまい、時間がかかってしまうことが課題としてあります。

自動で何か所か巡回してという場合に、ものが配置されており巡回できないとなった場合はすぐにあきらめて、スキップして次に行くということが実現したいと考えております。

**Reply by inabas-pfr** [reply]
こちらの機能ですが、アプリ操作とAPIともに早くて来月中旬にはリリースを予定しております。
検証の結果によってはリリースが数か月ずれる可能性もございますのでご了承いただけますと幸いです。

**Reply by inabas-pfr** [reply]
@yutaroha 
こちらの機能はv3.9.5でリリースされました。
[カチャカソフトウェアアップデートのお知らせ（Ver 3.9.5）](https://kachaka.zendesk.com/hc/ja/articles/11917387832335)
> [Pro] "目的地到着にかかる時間の上限"を設定できるようになりました

ご確認のほどよろしくお願いいたします。

**Reply by yutaroha** [reply]
@inabas-pfr 
ありがとうございます。無事動作いたしました。

目的地までの距離に関係なく一律の設定となっているため少し設定が難しいケースもあるかと感じました。

カチャカが各ウェイポイントに自動巡回し、計測するようなシステムを製作しています。一部のポイントに障害物となるものが置かれている場合にスキップしたいというニーズがあり、時間設定をしたいといううニーズがありました。

目的地までの距離は一定でなく可変のため、移動開始～終了までの時間で設定するのではなく、目的地に到着できずその場で立ち止まり立往生しているところからの時間が設定できるとうれしいと思いました。

目的地付近に障害物などが置かれている場合は、到着できず立ち止まり周りを見渡すような動作をして立往生してしまうので、単位時間あたりに移動量が少ない場合(立往生している場合)の時間経過のタイムアウトが別で設定可能であるとうれしいのではと思います。

ご参考になれば幸いです。

**Comment by atinfinity** [comment]
カチャカ正面に障害物があると、速度指令を与えても止まる（＝速度指令を受け付けない）安全機能が動いているという認識です。研究でカチャカを使ってPython API or ROS 2経由で動かす場合、カチャカが止まったのがこの安全機能由来なのかどうかを知りたいことがあります。現状のカチャカAPIのコードを読むとこの安全機能が作動しているかどうかのステータスは取れなさそうに見受けられますが、安全機能が作動しているかがAPIで取れる、もしくは安全機能を無効化ことができると嬉しいです。

**Comment by corgi0118** [comment]
Navigation2のthrough posesのように複数の通過ウェイポイントを指定できる移動指令APIを追加してほしいです。（通過ウェイポイントではカチャカは停止しないようにしてほしいです。）

**Reply by corgi0118** [reply]
また、このときのwaypointの通過状況（どのwaypointを通過したか、など）も取得できるAPIもほしいです

**Comment by Motohiro1192** [comment]
走行距離等を「日にち毎」に表示できないでしょうか？現在のアプリは「月ごと」に走行距離と移動した場所の情報が表示されていますが、この内容をカレンダー選択式で日にち毎に情報表示して欲しいです。1日毎の利用状況の確認がしたく。宜しくお願い致します。

**Comment by kazuonagae-jizaie** [comment]
お世話になっております。
カチャカのマイクで拾った音を、そのまま伝えられませんか？
カチャカのそばに居る人と、遠隔操作（インターネットの先の人）がカチャカを介在させて会話するイメージです。
遠隔側から声でカチャカを操作できると更に良いですね。
